consul/acl_replication.go

```diff
@@ -0,0 +1,351 @@
+package consul
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/consul/consul/structs"
+	"github.com/hashicorp/consul/lib"
+)
+
+// aclIDSorter is used to make sure a given list of ACLs is sorted by token ID.
+// This should always be true, but since this is crucial for correctness and we
+// are accepting input from another server, we sort to make sure.
+type aclIDSorter struct {
+	acls structs.ACLs
+}
+
+// See sort.Interface.
+func (a *aclIDSorter) Len() int {
+	return len(a.acls)
+}
+
+// See sort.Interface.
+func (a *aclIDSorter) Swap(i, j int) {
+	a.acls[i], a.acls[j] = a.acls[j], a.acls[i]
+}
+
+// See sort.Interface.
+func (a *aclIDSorter) Less(i, j int) bool {
+	return a.acls[i].ID < a.acls[j].ID
+}
+
+// aclIterator simplifies the algorithm below by providing a basic iterator that
+// moves through a list of ACLs and returns nil when it's exhausted.
+type aclIterator struct {
```

Comment:
Yeah that'll work! I didn't want to add it to the struct since it's pretty specific to this use case.